glyphic-co/textblock
# Enforce linting and formatting code on every commit

### Background

This PR is an addendum to PR #25 (specifically commit fef3bea925adc87119d90bffefe9d647c15c9dcb, which integrates ESLint and Prettier). I should have included this then, but I only learned about enforcing linting afterwards!

ESLint and Prettier are great tools for catching coding errors and enforcing a standard coding style, but of course they're useless unless you remember to run them (if you don't configure your editor to run them automatically).

### What This PR Does

This PR configures the project so you (and anyone contributing to your project) don't have to remember to lint. It runs `eslint --fix` (catching errors and making any possible automatic fixes) every time you try to commit code — even if you use a git GUI (like [Github Desktop](https://desktop.github.com), as I often do).

### How It Works

This PR sets up [husky](https://github.com/typicode/husky) (a widely used tool that makes it easy to hook into git commands) and [lint-staged](https://github.com/okonet/lint-staged) (also widely used, and which, as you can guess, lints staged files), so that no code with JS errors (as defined in .eslintrc.js) can be committed to the repo.

lint-staged runs `eslint --fix` on commit, which also formats your code, since Prettier is configured to run when ESLint runs. If you're not happy with Prettier's formatting, you can [configure options](https://prettier.io/docs/en/options.html) in .prettierrc.js.

### How To Test It

1. This PR adds the husky and lint-staged packages, so after you check it out locally, make sure to run `npm install`.
1. Add a line to `src/textblock.js`, like `var unusedVar = 'This is an unused variable.'`
1. Try committing the change. 

You should see an ESLint error like `error  'unusedVar' is assigned a value but never used`.
